pkg/agent/cniserver/container_configuration_windows.go

```diff
@@ -0,0 +1,373 @@
+// addresses and routes to the interface. It then sends a gratuitous ARP to the network.
+func attachContainerLink(ep *endpoint, containerID, sandbox, containerIFDev string) (*current.Interface, error) {
+	var found bool
+	for _, c := range ep.containers {
+		if c == containerID {
+			found = true
+		}
+	}
+
+	if !found {
+		if err := hcsshim.HotAttachEndpoint(containerID, ep.hnsEP.Id); err != nil {
+			return nil, err
+		}
+		ep.containers = append(ep.containers, containerID)
+	}
+	containerIface := &current.Interface{
+		Name:    strings.Join([]string{ep.hnsEP.Name, containerIFDev}, "_"),
+		Mac:     ep.hnsEP.MacAddress,
+		Sandbox: sandbox,
+	}
+	return containerIface, nil
+}
+
+// advertiseContainerAddr sends GARP packets in another goroutine with 50ms interval. It's because Openflow entries are
+// installed async, and the gratuitous ARP could be sent out after the Openflow entries are installed. Using another
+// goroutine to ensure the processing of CNI ADD request is not blocked.
+func (pc *podConfigurator) advertiseContainerAddr(containerNetNS string, containerIfaceName string, result *current.Result) error {
+	klog.V(2).Info("Send gratuitous ARP from container interface is not supported on Windows, return nil")
+	return nil
+}
+
+// removeContainerLink removes the HNSEndpoint attached on the Pod.
+func (pc *podConfigurator) removeContainerLink(containerID, epName string) error {
+	ep, found := pc.getEndpoint(epName)
+	if !found {
+		return nil
+	}
+	return pc.removeHnsEndpoint(ep, containerID)
+}
+
+// removeHnsEndpoint removes the HNSEndpoint from HNS and local cache.
+func (pc *podConfigurator) removeHnsEndpoint(endpoint *endpoint, containerID string) error {
+	epName := endpoint.hnsEP.Name
+	// Remove HNSEndpoint.
+	_, err := endpoint.hnsEP.Delete()
+	if err != nil {
+		if !strings.Contains(err.Error(), notFoundHNSEndpoint) {
+			klog.Errorf("Failed to delete container interface %s: %v", containerID, err)
+			return err
+		}
+	}
+	// Delete HNSEndpoint from local cache.
+	pc.delEndpoint(epName)
+	return nil
+}
+
+func parseContainerIfaceFromResults(cfgArgs *cnipb.CniCmdArgs, prevResult *current.Result) *current.Interface {
+	for _, intf := range prevResult.Interfaces {
+		if strings.HasSuffix(intf.Name, cfgArgs.Ifname) {
+			return intf
+		}
+	}
+	return nil
+}
+
+// checkContainerInterface finds the virtual interface of the container, and compares the network configurations with
+// the previous result.
+func (pc *podConfigurator) checkContainerInterface(
+	sandboxID, containerID string,
+	containerIface *current.Interface,
+	containerIPs []*current.IPConfig,
+	containerRoutes []*cnitypes.Route) (*vethPair, error) {
+
+	// Check container sandbox configuration.
+	if sandboxID != containerIface.Sandbox {
+		klog.Errorf("Sandbox in prevResult %s doesn't match configured sandbox: %s",
+			containerIface.Sandbox, sandboxID)
+		return nil, fmt.Errorf("sandbox in prevResult %s doesn't match configured netns: %s",
```

Comment:
I suggest not to log and return the same error at the same time, otherwise finally we will see the error twice in logs.